File: src/main/scala/workshop/part3/SuperComputeRouter.scala
package workshop.part3

import akka.actor.SupervisorStrategy.Resume
import akka.actor.{Actor, ActorRef, OneForOneStrategy, Props}
import akka.event.Logging
import akka.routing._
import workshop.work.RiskyWork

import scala.concurrent.duration._
import scala.language.postfixOps

class SuperComputeRouter extends Actor {
  val log = Logging(context.system, this)

  val resizer = DefaultResizer(lowerBound = 5, upperBound = 20, messagesPerResize = 1)
  val router: ActorRef =
    context.actorOf(RoundRobinPool(10, Some(resizer)).props(Props[Worker]), "router1")

  override val supervisorStrategy =
    OneForOneStrategy(maxNrOfRetries = 10, withinTimeRange = 1 minute, loggingEnabled = false) {
      case _: Exception => Resume
    }

  def receive = {
    case w : RiskyWork =>
      router.tell(w, sender())
  }
}

File: src/main/java/workshop/part3/SuperComputeRouter.java
package workshop.part3;

import akka.actor.*;
import akka.japi.pf.ReceiveBuilder;
import akka.routing.DefaultResizer;
import akka.routing.RoundRobinPool;
import scala.PartialFunction;
import scala.concurrent.duration.Duration;
import scala.runtime.BoxedUnit;

import static akka.actor.SupervisorStrategy.resume;
import static java.util.concurrent.TimeUnit.MINUTES;
import static workshop.work.Work.RiskyWork;

public class SuperComputeRouter extends AbstractActor {

   private RoundRobinPool roundRobinPool = new RoundRobinPool(10)
           .withResizer(new DefaultResizer(5, 20));

    ActorRef router =
            getContext().actorOf(roundRobinPool.props(Props.create(Worker.class)), "router-1");

    private SupervisorStrategy strategy =
            new OneForOneStrategy(10, Duration.create(1, MINUTES),
                    t -> resume(), false);

    @Override
    public SupervisorStrategy supervisorStrategy() {
        return strategy;
    }

    @Override
    public PartialFunction<Object, BoxedUnit> receive() {
        return ReceiveBuilder
                .match(RiskyWork.class, w -> router.tell(w, sender()))
                .build();
    }
}
